### PR TITLE
gh-94912: Added marker for non-standard coroutine function detection

### PR DESCRIPTION
**Pull request by carltongibson**
This is ref the discussion on https://github.com/python/cpython/issues/94912 to add an official way of marking callables as coroutine functions where they would not otherwise be detected. 

cc. @gvanrossum @andrewgodwin I've just had time this afternoon to block off some first tests, and docs. I wanted to break ground so we keep it moving. It's very consciously **drafts**: please make suggestions of how you think it should go. 

@gvanrossum Can I ask for a pointer on how I'm meant to set `co_flags` for the `_has_code_flag()` check? 🤔 

https://github.com/python/cpython/blob/c43714fbcdb9bb0927872d6ebf5697edd2e2a1e9/Lib/inspect.py#L376-L385

Thanks! 

First PR on CPython, so likely it's wrong 😊

* Clone and Build as per https://devguide.python.org. 
* Run new just tests with `% ./python.exe -m unittest test.test_inspect.TestPredicates`
* Do we need to create a new ticket, or is #94912 OK? 

* Issue: gh-94912

### CODE REVIEW

**Comment by sobolevn** [comment]
@carltongibson happy to see you here! Feel free to ask any questions you have: I am happy to help with what I can :)

1. The simplest way of adding news is by using https://blurb-it.herokuapp.com/ 
2. Running tests is better with `./python.exe -m test -v test_inspect`

**Comment by carltongibson** [comment]
Hi @sobolevn — thanks for your pointer very helpful. 

I've pushed two commits, making the tests pass: 

* https://github.com/python/cpython/pull/99247/commits/ab422f7ec02a55c93b620670c415e76279247192 uses the `__code__.replace()` technique, and looks good. (I need to move this to a decorator — I'll work on this next.)
* https://github.com/python/cpython/pull/99247/commits/8c6429ec008521e9e8cb8110c8221e308442835d adjusts the iscoroutinefunction() function to check the extra case of a class with an `async def __call__()`. I went this way because I wasn't able to get the `__code__` approach working. (`AttributeError: type object 'Cl' has no attribute '__code__'. ...`) 

I wanted to push that to see if I could get feedback. Thanks. 

> I don't know anything about its performance

Certainly in our usage, this marker would be applied once at startup I think. If one were doing this in a hot-loop, I'd think the _just use `async def`_ response would become dominant. 🤔

File: Lib/inspect.py
Comment (sobolevn) [inline Lib/inspect.py]: Right now the function is called `iscoroutinefunction`. But, if applied to a callable object with `__call__` marked as `@markcoroutine` it will return `True`.

I might be wrong, but this does not seem like a matching semantics.

Other similar functions do not do that.

File: Lib/test/test_inspect.py
Comment (sobolevn) [inline Lib/test/test_inspect.py]: In the future don't forget to test objects with `.__func__` set, because of how `_has_code_flag` works :)

Comment (carltongibson) [reply Lib/test/test_inspect.py]: I haven't followed exactly what you mean here. Sorry. 🤔

File: Lib/inspect.py
Comment (carltongibson) [reply Lib/inspect.py]: @sobolevn — OK, I might need some direction then on the way forward. 🤔 

Despite the name — `iscoroutinefunction` — class instances with either an `async def __call__` or a `def __call__` where that has been marked with the new decorator — where we assume the writer knew what they were doing —  **should** return `True` here.

File: Lib/test/test_inspect.py
Comment (sobolevn) [reply Lib/test/test_inspect.py]: Oh, sorry. I've meant that objects like `staticmethod` and `classmethod` have `__func__` property set, where the original function is stored. `inspect` generally supports this pattern and it should be tested in this case as well :)

**Comment by carltongibson** [comment]
OK, I'm marking Ready for Review, since this would resolve our need from the discussion in #99247 to be able to continue to identify sync functions returning awaitables, without calling them to see that. The behaviour is as I'd expect. 

* I'm quite sure that it's not perfect, but I need guidance on how you want it to be. Quite happy to make adjustments. 
* There were a couple of comments from @sobolevn that I didn't follow. Hopefully that additional tests clarify what's needed. 

Thanks!

File: Doc/library/inspect.rst
Comment (sobolevn) [inline Doc/library/inspect.rst]:
```suggestion
.. function:: markcoroutinefunction(func)
```

File: Lib/test/test_inspect.py
Comment (carltongibson) [reply Lib/test/test_inspect.py]: @sobolevn — OK, let me have a play and see. Thanks!

File: Lib/test/test_inspect.py
Comment (sobolevn) [inline Lib/test/test_inspect.py:208]: Let's also test `lambda`?

**Review by sobolevn — COMMENTED** [review]
Sidenote: `asyncio.iscoroutinefunction` will also be affected by this change.

**Comment by gvanrossum** [comment]
Let me know when you both feel this is ready to go in.

**Comment by carltongibson** [comment]
Thanks @gvanrossum. 

I'm going to add the extra test cases, and then (from my POV) it's a _Any changes?_ question. 

> Sidenote: asyncio.iscoroutinefunction will also be affected by this change.

That's being made an alias to the `inspect` version as part of gh-94912 IIUC — so are additional changes (doc maybe) needed as part of **this** PR? 🤔

Thanks for the guidance both.

**Comment by carltongibson** [comment]
Hi @sobolevn and @gvanrossum — thanks for your patience. I've added the additional tests, and adjusted the docs as suggested. I can rebase/squash/etc as you need, but I think it's ready for you to look at. (Happy to adjust as you think!)

For my POV this would allow removal of the `asyncio.iscoroutinefunction` (and the `_is_coroutine` marker attribute): we'd introduce a compatibility shim until 3.12 was the lowest supported version, but that's easily enough done. 
Thanks again. 


//cc @andrewgodwin.

**Review by sobolevn — COMMENTED** [review]
Thank you! I hope your first PR to CPython was a pleasant experience 😉 

I have just a single main concern: that `inspect.iscoroutinefunction` no longer does what is documented.

Citing:
> Return True if the object is a [coroutine function](https://docs.python.org/3/glossary.html#term-coroutine-function) (a function defined with an [async def](https://docs.python.org/3/reference/compound_stmts.html#async-def) syntax).

Now it can return `True` for instances with async `__call__` methods.
This might backfire for users that trust this check right now.

I suggest to (but, this is just *my* opinion, I don't know what others think of it):
- `markcoroutinefunction` is a great thing to have 👍 
- But, maybe we can delay making changes to `iscoroutinefunction` to a separate PR? I see other options like creating new `isdefiningcoroutine` (or any similar name) function.

Or we can change the docs and list all special cases.

**Comment by carltongibson** [comment]
Thanks @sobolevn, yes it's been fine. 

Good point on the docs. Whichever way we go, a small tweak there. 👀

Some thoughts to your concern:

If I followed correctly, the issue is (only) with the `Cl2` example, with the `async def __call__`. 

Given there's no existing test covering this, I think it's similar to how support for `partial` instances was added here in Python 3.8. It's a case we do want True for, and should add (and doc). 

I note the Starlette framework have the same detection for such a class in place. (On phone so don't have link) So it's not just Django that's needing this behaviour. 

The name `iscoroutinefunction` gets stretched a little... I'm not sure a second function would carry its weight. 

🤔

File: Misc/NEWS.d/next/Library/2022-11-17-10-02-18.gh-issue-94912.G2aa-E.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2022-11-17-10-02-18.gh-issue-94912.G2aa-E.rst]: I'd rephrase this using present tense (like most other news entries) and using ```:func:`inspect.markcoroutinefunction` ```.
```suggestion
Add :func:`inspect.markcoroutinefunction` decorator which manually marks
a function as a coroutine for the benefit of :func:`iscoroutinefunction`.
```

**Comment by gvanrossum** [comment]
I think it's better not to add a new function, it would require lots of places to be updated to call the new function. Setting the marker is meant to avoid that need. But I'm okay with a slight doc update. I'll hold off approving (although it LGTM) for now.

**Comment by carltongibson** [comment]
Hi all. Thank you for your patience. I've adjusted the docs. Any input on phrasing welcome.

File: Doc/library/inspect.rst
Comment (gvanrossum) [inline Doc/library/inspect.rst]: I would add an explicit mention of sync functions marked with `@markcoroutinefunction` here (and in the versionchanged block too).

File: Doc/library/inspect.rst
Comment (gvanrossum) [inline Doc/library/inspect.rst]: This is a bit confusing because of the comma. Suggestion:

```suggestion
   This may be of use for sync functions that return a :term:`coroutine`,
   if the function is passed to an API that requires :func:`iscoroutinefunction`.

   When possible, using an :keyword:`async def` function is preferred.
   Also acceptable is calling the function
   and testing the return with :func:`iscoroutine`.
```

Feel free to reword this to your pleasing.

**Review by gvanrossum — CHANGES_REQUESTED** [review]
(PS. Don't force-push, that makes the review process harder.)

File: Doc/library/inspect.rst
Comment (carltongibson) [reply Doc/library/inspect.rst]: Thanks @gvanrossum. 

I went with your suggestion here. 

One option I looked at was: 

```
   When possible, using an :keyword:`async def` function is preferred.
   Alternatively, consider calling the function and testing the return with
   :func:`iscoroutine`, if feasible. 
```

The point was to capture your initial concerns from the discussions on the issue, and those linked from there. I took that to be ≈_"Do these other things if you can"_. I'm happy to defer to you on whether that's valuable (I guess it is!) and exact wording.

**Comment by carltongibson** [comment]
I have made the requested changes; please review again

> (PS. Don't force-push, that makes the review process harder.)

Sorry, yes. It was an instant `!fixup` on the tip commit, so I hoped to correct that to reduce noise. I'll not do so next time. Thanks.

**Review by gvanrossum — APPROVED** [review]
Yes! I will merge it for you.

**Review by kumaraditya303 — DISMISSED** [review]
Please add a what's new entry for this.

**Comment by gvanrossum** [comment]
Agreed with Kumar, this needs a little "what's new in 3.12" bit.

Note: in the past we didn't require that, some release manager would write those that based on the change logs. But it's easier on the release manager if individual contributors write their own "what's new" bits. Let us know if you need help with this.

**Comment by TeamSpen210** [comment]
Is it safe to set `CO_COROUTINE` on a code object which isn't actually compiled as `async def`? That would cause even the interpreter loop to think that the function actually is a coroutine, and do things like [raise unawaited coroutine](https://github.com/python/cpython/blob/135ec7cefbaffd516b77362ad2b2ad1025af462e/Objects/genobject.c#L92) warnings if the function was a generator.

**Comment by carltongibson** [comment]
> this needs a little "what's new in 3.12" bit.

No problem. I'll add that. 

> ... do things like [raise unawaited coroutine](https://github.com/python/cpython/blob/135ec7cefbaffd516b77362ad2b2ad1025af462e/Objects/genobject.c#L92) warnings... 

I think that's desired behaviour 🤔 If I mark a sync function with the new decorator, it would be a programming error to not `await` its return value, so reporting that would be desired. 

Not sure what to make of `if the function was a generator` — I can't immediately see the problem case that you'd pass to the new decorator. I guess it's possible to add guards if there are likely ones? 🤔

**Comment by gvanrossum** [comment]
Whoa. I must've not paid attention when I approved this. :-(

I don't believe this should be implemented by setting the `CO_COROUTINE` flag in `code.co_flags`. That flag has meaning for the interpreter and I don't think the bytecode is a regular function is compatible with that flag.

I had understood when we first decided to do this that the mark decorator would set a simple flag attribute on the _function_ object which `iscoroutinefunction` would then check. This is how `asyncio.iscoroutinefunction` works.

Am I missing something?

**Comment by carltongibson** [comment]
Ok, so I should rewrite it setting an attribute rather than using the code flags? (Can do)

**Comment by gvanrossum** [comment]
> Ok, so I should rewrite it setting an attribute rather than using the code flags? (Can do)

Yeah, since you don't seem to recall a discussion leading to the conclusion that we could use the code flags, I think that would be safer.

**Comment by carltongibson** [comment]
I have made the requested changes; please review again

* Reworked to use a `_is_coroutine` marker. 
* Added "What's New…" entry.

File: Lib/inspect.py
Comment (carltongibson) [inline Lib/inspect.py]: For the check, we need to check plain functions and the `__func__` cases, **and** the `__call__` implementation for class instances. (Tests cover that: given that functions have a `__call__` we need to be careful not to just check the one case.) 

I looked at various ways of combining this into a single pass, but none that I found were particularly clear. Happy to take a suggestion, but I don't see this as being performance sensitive. 

The implementation here is more sophisticated that the older `asyncio` one…

https://github.com/python/cpython/blob/3.11/Lib/asyncio/coroutines.py#L17-L24

… but (as per the added test cases) various extra (seemingly legitimate) cases are now covered. (So 👍 I guess)

Comment (gvanrossum) [reply Lib/inspect.py]: I'm thinking about this, but being distracted by other stuff. I'll try to get to it later this week.

Comment (carltongibson) [reply Lib/inspect.py]: Hey @gvanrossum — thanks. 

I thought one maybe neater re-write looks like this: 

```diff
diff --git a/Lib/inspect.py b/Lib/inspect.py
index 883d0b02bf..3b9bd14a38 100644
--- a/Lib/inspect.py
+++ b/Lib/inspect.py
@@ -410,12 +410,13 @@ def iscoroutinefunction(obj):
     Coroutine functions are normally defined with "async def" syntax, but may
     be marked via markcoroutinefunction.
     """
-    func = getattr(obj, "__func__", obj)
-    if getattr(func, "_is_coroutine", None) is _is_coroutine:
-        return True
-
-    if not isclass(obj) and callable(obj) and getattr(obj.__call__, "_is_coroutine", None) is _is_coroutine:
-        return True
+    if not isclass(obj) and callable(obj):
+        # Test both the function and the __call__ implementation for the
+        # _is_coroutine marker.
+        f = getattr(getattr(obj, "__func__", obj), "_is_coroutine", None)
+        c = getattr(obj.__call__, "_is_coroutine", None)
+        if f is _is_coroutine or c is _is_coroutine:
+            return True
 
     return _has_code_flag(obj, CO_COROUTINE) or (
         not isclass(obj) and callable(obj) and _has_code_flag(obj.__call__, CO_COROUTINE)
```

Comment (gvanrossum) [reply Lib/inspect.py]: So perhaps a confusing aspect is that _everything_ that's callable has a `__call__` method. Also the presence of `__func__` indicates it's a class or instance method.

At least, those things seem to be obfuscating what's going on a bit.

Also, looking at the code for `_has_code_flag()`, I wonder if we don't need the same logic for checking the `_is_coroutine` flag.

What would happen if we wrote a helper function like this?
```
def _has_coroutine_mark(f):
    while ismethod(f):
        f = f.__func__
    f = functools._unwrap_partial(f)
    if not (isfunction(f) or _signature_is_functionlike(f)):
        return False
    return getattr(f, "_is_coroutine", None) is _is_coroutine
```

(FWIW I wonder if the variable `_is_coroutine` shouldn't be renamed `_is_coroutine_marker`.)

Now, I'm not entirely sure why you're testing for `__call__`, when none of the other `is<something>function()` predicates test for it. Maybe things would become simpler if we removed that feature?

We could then define our function like this:
```
def iscoroutinefunction(obj):
    return _has_code_flag(obj, CO_COROUTINE) or _has_coroutine_mark(obj)
```

Comment (carltongibson) [reply Lib/inspect.py]: > Now, I'm not entirely sure why you're testing for __call__

OK, so the issue is we have classes like this: 

```
        class Cl:
            async def __call__(self):
                pass
```

... where we need **instances** to be identified as async, and `iscoroutinefunction` doesn't (currently) pick it up. 

Note that this is an actual, already-out-there, need. e.g. [the Starlette framework has this exact check to work around this limitation](https://github.com/encode/starlette/blob/88e9fc1411f6bd79131afa4a7d2f4dc576c8bf04/starlette/_utils.py#L6-L12). `asgiref` works around this limitation in its usage by marking the instance itself with the `_is_coroutine` marker. 

Either way, I think it is desirable to return `True` for class instance of classes such as `Cl` with an `async def __call__()`

So that's the reason for adding the additional `... _has_code_flag(obj.__call__, ...)` check. 

In this PR, given that we were adding `@staticmethod` and `@classmethod` and so on, it seemed that we also should cover this kind of case: 

```
        class Cl2:
            @inspect.markcoroutinefunction
            def __call__(self):
                pass
```

Which leads to needing similar in the `_is_coroutine_marker` block. 

The latter case is somewhat hypothetical... — it comes from trying to cover all the cases here, rather than from real-use™. To that extent I'd be happy to drop it, but I guess if we do there's an issue in X months time, saying "I need to do this". 🤔

I don't know the correct thing to say. (?)

> (FWIW I wonder if the variable _is_coroutine shouldn't be renamed _is_coroutine_marker.)

Yep OK. +1

> What would happen if we wrote a helper function like this?

Let me have a read. 

Thanks so much for your thought and input on this.

Comment (gvanrossum) [reply Lib/inspect.py]: Hum, I still feel that you're sneaking in a new feature under the radar. :-) What this PR is supposed to fix is _just_ to add the ability to mark a sync function as "morally async". IOW the behavior of
```
class C:
    @inspect.markcoroutinefunction
    def __call__(self): ...
print(inspect.iscoroutinefunction(C())
```
should be the same as
```
class C:
    async def __call__(self): ...
print(inspect.iscoroutinefunction(C())
```
and since the latter prints `False`, I don't think _this_ PR has any business changing the output to `True`.

If you think the functionality used by Starlette deserves to be included in inspect, you should probably open a new issue for that first so it can be discussed properly. But I personally think this is overreaching (however, I am repeating myself).

Have you considered adding `def _has_coroutine_mark(f):` as I suggested?

(Sorry to be such a pain. But once we let this in there's no way back, so I prefer to have the minimal necessary functionality only. Other core devs may differ though.)

**Comment by carltongibson** [comment]
Hi @gvanrossum — I've updated as per your advice. Please let me know what you think. 

> Sorry to be such a pain. But once we let this in there's no way back, so I prefer to have the minimal necessary functionality only...

Absolutely, I understand that. 

> ...so it can be discussed properly...

That's slightly frustrating (🙂), as I thought that was exactly the conversation we'd had leading to your https://github.com/python/cpython/pull/99247#issuecomment-1325971890 above: 

> I think it's better not to add a new function ... 

The issue is that `inspect.iscoroutinefunction` does not capture all the cases it needs to. The `asyncio` version doesn't either, but this surfaces because the proposal to remove that version makes the deficiency (breakingly) worse. 

The two missing cases are: 

* Sync functions that return coroutine objects. 
* Class instances with `async def __call__`. 

If we proceed here without handling the `__call__` case we leave that broken (forcing clients to implement their own different shims for each framework) I can put that in a separate issue or PR or commit but as I'm reading you, you wouldn't accept that anyway? 

Beyond those added here, there are no tests for the classes defining `async def __call__` methods with `iscoroutinefunction`. But as a consumer of `iscoroutinefunction` I (clearly) want `True` in such cases (and the hand-marked equivalents.)  Fixing this seems no different that fixing `functools.partial` support in Python 3.8. 

> …however, I am repeating myself…

So, yes, we're going round in circles. 🙂 

Ultimately, I have to go with your decision (no problem). If you're not convinced then there's little left to say. Please advise if you think I should re-add `__call__` support here, or open a new ticket, or leave it. 

Thanks for your patience and guidance on this PR, and all round work beyond that. 🎁

**Comment by carltongibson** [comment]
@gvanrossum I think this is now as you want it. Please let me know if there's further to do here. Thanks.

**Review by gvanrossum — APPROVED** [review]
Perfect!

**Comment by gvanrossum** [comment]
Thanks a lot for this improvement. Sorry it took us all a while to settle on the right way to do it.

**Comment by carltongibson** [comment]
Thanks @gvanrossum. No worries! Better to be sure. Thanks again for your help!

**Pull request merged**
